src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -1781,6 +1781,40 @@ bool ReplicatedMergeTreeMergePredicate::operator()(
     return true;
 }
 
+bool ReplicatedMergeTreeMergePredicate::canMergeSinglePart(const MergeTreeData::DataPartPtr & part, String * out_reason) const
+{
+    LOG_FATAL(&Poco::Logger::get("ReplicatedMergeTreeMergePredicate::operator()"), "begin");
+
+    if (part->name == inprogress_quorum_part)
+    {
+        LOG_FATAL(&Poco::Logger::get("ReplicatedMergeTreeMergePredicate"), "operator()");
+        if (out_reason)
+            *out_reason = "Quorum insert for part " + part->name + " is currently in progress";
+        return false;
+    }
+
+    if (prev_virtual_parts.getContainingPart(part->info).empty())
+    {
+        if (out_reason)
+            *out_reason = "Entry for part " + part->name + " hasn't been read from the replication log yet";
+        return false;
+    }
+
+    std::lock_guard<std::mutex> lock(queue.state_mutex);
+
+    /// We look for containing parts in queue.virtual_parts (and not in prev_virtual_parts) because queue.virtual_parts is newer
+    /// and it is guaranteed that it will contain all merges assigned before this object is constructed.
+    String containing_part = queue.virtual_parts.getContainingPart(part->info);
```

Comment:
Does it mean, that previously we could assign multiple equal merges for the first part in partition? With the same output name?

Reply:
How I think previosly we always assign a merge for the first part. That's why it will be assigned as many merges as how many attemts was.

Reply:
I've found such a discussion https://github.com/dropbox/djinni/issues/36